README.md
# Riot Lol API

This module lets you query the Riot API for LeagueOfLegends data.

You'll need a developer key from https://developer.riotgames.com.

This module was developed for people that need to poll the Riot API with a *very high* throughput (with peaks above the standard production rate limit of 300 calls / second / region).

If you don't need this kind of performance, you'll probably be better with other modules -- have a look at [lol-js](https://www.npmjs.com/package/lol-js) for instance :)

## Installation
```
npm install --save riot-lol-api
```

## Usage
```js
var RiotRequest = require('riot-lol-api');

var riotRequest = new RiotRequest('my_api_key');

// 'summoner' is a string to identify the method being used currently
// See note about rate-limiting in the README.
// Also see https://developer.riotgames.com/rate-limiting.html#method-headers
riotRequest.request('euw1', 'summoner', '/lol/summoner/v3/summoners/by-name/graphistos', function(err, data) {});
```

The library will take care of rate limiting and automatically retry on 500 and 503.

It will also maintain a very high request concurrency, dynamically updating concurrency to ensure you remain a good citizen and don't get blacklisted.

Ensure that your network adapter can deal with the traffic!
If necessary, you can distribute the library across multiple servers -- I'm currently using it with a production key distributed on 4 servers sending > 35 millions calls a day.

## Caching
The second argument in the constructor lets you define a cache object. This object should expose two keys, `get` and `set`. The default implementation does no caching:

```js
var cache = {
  get: function(region, endpoint, cb) {
    // Try to read from cache,
    // Return cb(null, data) if data is already available in your cache.
    // If it's a cache-miss, you still need to call cb(null, null) for the request to proceed.
    // Do not just call cb(null)!
    cb(null, null);
  },
  set: function(region, endpoint, cacheStrategy, data) {
    // Use this function to store `data`, which is the result of the API call to `endpoint` on `region`.
  }
};
```

`cacheStrategy` is a value over which you have total control when you call `.request()`:


```js
riotRequest.request('euw1', 'summoner', '/lol/summoner/v3/summoners/by-name/graphistos', YOUR_CACHE_STRATEGY, function(err, data) {});
```

When unspecified, `cacheStrategy` will default to `false`, and cache won't be used.
If the value is not falsy, the cache will be used and the value will be forwarded to you (in your `.set` cache method). The most common use case would be to send how long you want to store the data in cache, but this is completely up to you.

You may want to use a package like `lru-cache` to help you with caching -- note that you can plug any system you want (Redis, Riak, file system), just ensure you call `cb(null, data)`. If you send an error in the first argument, the library will forward this error directly to the callback specified in `.request()`.

You'll notice that the `set()` function has no callback, this is intentional. You can start async operations from here, but the system won't wait for your operation to complete before moving on to other requests.

In some situations, the `get()` function might be called more than once per endpoint. For performance, when a request is queued, it is checked instantly if it's in cache: if it isn't, it's added in a queue, and when the worker start that task he will ask the cache again in case the same request was already queued and has since then been cached.

## Rate limiting
The Riot API rate limiting is complex -- see https://developer.riotgames.com/rate-limiting.html for all the nitty gritty.

This library abstracts most of it away, automatically reading the headers values and adjusting this behavior to ensure your key doesn't get blacklisted.

However, when you call `.request`, you need to specifiy a string to identify the method currently being used.

A list of all the buckets is available in https://developer.riotgames.com/rate-limiting.html#method-headers, but the TL;DR is that for every type of request you send, you should have some kind of tag: for instance, all requests for recent games can be tagged with "recent-games" (the second parameter to `.request(region, tag, endpoint)`. `riot-lol-api` will then ensure that all rate limits (both for the app and for the method) are respected per region.

If the above paragraph didn't make any sense, go and check out the official Riot link above and then come back to this section ;)

Here is a sample code excerpt: 

```js
riotRequest.request('euw1', 'summoner', '/lol/summoner/v3/summoners/by-name/graphistos', function(err, data) {});
riotRequest.request('euw1', 'champion-mastery', '/lol/champion-mastery/v3/champion-masteries/by-summoner/4203456', function(err, data) {});
riotRequest.request('euw1', 'league', '/lol/league/v3/positions/by-summoner/4203456', function(err, data) {});
```

## Logging
The library use `debug` for logging. To see logs, set this environment variable: `DEBUG=riot-lol-api:*`.

## Errors
Errors when trying to read the cache are forwarded directly to the requester.

HTTP errrors on the Riot API side will expose three properties:

* `.statusCode` containing the return code from the API (the most common one is 503. Note that the library is retrying by default all 5XX errors, so if you see it in your code it means that the error happened twice)
* `riotInternal` a flag set to true to help you distinguish network errors (fairly common) from more standard errors (e.g. from your cache)
* `extra`, an object exposing details about the request: endpoint, region, status code, whether the failure is due to a timeout... You may want to send this object directly to you error monitoring system.

Please remember that the library will automatically retry once when it receives a 500 and 503.

## Dealing with regions and platforms
For convenience, the library exposes a function `getPlatformFromRegion()` that takes a region as parameter (e.g "euw") and returns the associaed platform to use with the Riot API ("EUW1"). This can be useful for building URLs.

Additionally, there is also a `.REGIONS` property with an array of all valid Riot regions lowercased.


CHANGELOG.md
See https://github.com/Neamar/riot-lol-api/releases for releases details.


test/index.js
"use strict";

var assert = require("assert");
var async = require("async");
var nock = require("nock");

var RiotRequest = require('../lib/index.js');

describe("Riot queue", function() {
  describe("Constructor", function() {
    it("should throw when missing Riot api key", function() {
      assert.throws(
        function() {
          /* jshint -W031 */
          new RiotRequest();
        },
        /missing riot api/i);
    });

    it("should throw when using invalid cache", function() {
      assert.throws(
        function() {
          /* jshint -W031 */
          new RiotRequest(123, [1]);
        },
        /invalid cache object/i);
    });
  });

  describe("Requester without cache", function() {
    var riotRequest = new RiotRequest("fake_key");

    it("should return results on valid reply from Riot's server", function(done) {
      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply({}, {ok: true});

      riotRequest.request('EUW', 'test', '/fake', false, function(err, res) {
        if(err) {
          return done(err);
        }

        assert.equal(res.ok, true);
        done();
      });
    });

    it("should retry automatically after a 500", function(done) {
      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(500, {ok: false});

      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(200, {ok: true});

      riotRequest.request('EUW', 'test', '/fake', false, function(err, res) {
        if(err) {
          return done(err);
        }

        assert.equal(res.ok, true);
        done();
      });
    });

    it("should fail after a second 500", function(done) {
      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(500, {ok: false});

      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(500, {ok: true});

      riotRequest.request('EUW', 'test', '/fake', false, function(err) {
        if(!err) {
          return done(new Error("Expected an error to occur."));
        }

        assert.equal(err.statusCode, 500);
        assert.equal(err.riotInternal, true);
        assert.equal(err.extra.endpoint, '/fake');
        done();
      });
    });

    it("should retry automatically after a 429", function(done) {
      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(429, {ok: false}, {'retry-after': '0.01'});

      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(200, {ok: true});

      riotRequest.request('EUW', 'test', '/fake', false, function(err, res) {
        if(err) {
          return done(err);
        }

        assert.equal(res.ok, true);
        done();
      });
    });

    describe("Rate limiting", function() {
      it("should honor rate limits", function(done) {
        // Only one concurrent request at a time
        var riotRequest = new RiotRequest("fake_key");

        nock('https://euw.api.riotgames.com')
          .get('/fake')
          .query(true)
          .reply(200, {ok: "bar"}, {
            'x-app-rate-limit-count': '1:1',
            'x-app-rate-limit': '1:1',
            'x-method-rate-limit-count': '1:1',
            'x-method-rate-limit': '1:1'
          });

        riotRequest.request('EUW', 'test', '/fake', false, function(err, res) {
          assert.ifError(err);

          assert.equal(res.ok, "bar");

          assert.equal(riotRequest.requestQueues.euwtest.concurrency, 1);
          done();
        });
      });

      it("should honor app rate limits", function(done) {
        var riotRequest = new RiotRequest("fake_key");

        nock('https://euw.api.riotgames.com')
          .get('/fake')
          .query(true)
          .reply(200, {ok: "bar"}, {
            'x-app-rate-limit-count': '3:1',
            'x-app-rate-limit': '10:1',
            'x-method-rate-limit-count': '1:1',
            'x-method-rate-limit': '100:1'
          });

        riotRequest.request('EUW', 'test', '/fake', false, function(err) {
          assert.ifError(err);

          // 10 - 3
          assert.equal(riotRequest.requestQueues.euwtest.concurrency, 7);
          done();
        });
      });

      it("should honor method rate limits", function(done) {
        var riotRequest = new RiotRequest("fake_key");

        nock('https://euw.api.riotgames.com')
          .get('/fake')
          .query(true)
          .reply(200, {ok: "bar"}, {
            'x-app-rate-limit-count': '1:1',
            'x-app-rate-limit': '100:1',
            'x-method-rate-limit-count': '5:1',
            'x-method-rate-limit': '30:1'
          });

        riotRequest.request('EUW', 'test', '/fake', false, function(err) {
          assert.ifError(err);

          // 30 - 5
          assert.equal(riotRequest.requestQueues.euwtest.concurrency, 25);
          done();
        });
      });

      it("should honor secondary rate limits", function(done) {
        var riotRequest = new RiotRequest("fake_key");

        nock('https://euw.api.riotgames.com')
          .get('/fake')
          .query(true)
          .reply(200, {ok: "bar"}, {
            'x-app-rate-limit-count': '1:10,1:600',
            'x-app-rate-limit': '1000:10,420000:600',
            'x-method-rate-limit-count': '1:10,1000:600',
            'x-method-rate-limit': '1000:10,1200:600'
          });

        riotRequest.request('EUW', 'test', '/fake', false, function(err) {
          assert.ifError(err);

          // 1200 - 1000
          assert.equal(riotRequest.requestQueues.euwtest.concurrency, 200);
          done();
        });
      });
    });

    // First call now has a concurrency of 1.
    it.skip("should allow for multiple calls in parallel", function(done) {
      var riotRequest = new RiotRequest("fake_key");

      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(200, {ok: "part1"});

      nock('https://euw.api.riotgames.com')
        .get('/fake')
        .query(true)
        .reply(200, {ok: "part2"});

      // Run in parallel
      async.parallel([
        function firstCall(cb) {
          riotRequest.request('EUW', 'test', '/fake', false, function(err, res) {
            if(err) {
              return cb(err);
            }

            assert.equal(res.ok, "part1");

            // Ensure the second calls fails it it hasn't already suceeded
            nock.cleanAll();
            nock('https://euw.api.riotgames.com')
              .get('/fake')
              .query(true)
              .reply(404, {ok: false});

            cb();
          });
        },
        function secondCall(cb) {
          riotRequest.request('EUW', 'test', '/fake', false, function(err, res) {
            if(err) {
              return cb(err);
            }

            assert.equal(res.ok, "part2");
            cb();
          });
        }
      ], done);
    });
  });

  describe("Requester with cache", function() {
    it("should let user specify its own cache function", function(done) {
      var riotRequest = new RiotRequest("fake", {
        get: function(region, endpoint, cb) {
          cb(null, "cached_value");
        },
        set: function(region, endpoint, cacheStrategy, data) {
          // jshint unused:false
          // Do nothing.
        }
      });

      riotRequest.request('EUW', 'test', '/cacheable', 150, function(err, data) {
        assert.ifError(err);
        assert.equal(data, "cached_value");

        done();
      });
    });

    it("should call the setter function on the cache object", function(done) {
      var defaultPayload = {ok: true};
      nock('https://euw.api.riotgames.com')
        .get('/cacheable')
        .query(true)
        .reply({}, defaultPayload);

      var requiredCacheStrategy = 150;
      var riotRequest = new RiotRequest("fake", {
        get: function(region, endpoint, cb) {
          cb(null, null);
        },
        set: function(region, endpoint, cacheStrategy, data) {
          assert.deepEqual(data, defaultPayload);
          assert.equal(requiredCacheStrategy, cacheStrategy);
          process.nextTick(done);
        }
      });

      riotRequest.request('EUW', 'test', '/cacheable', requiredCacheStrategy, function(err) {
        assert.ifError(err);
      });
    });

    it("should not call the setter function when reading from cache", function(done) {
      var requiredCacheStrategy = 150;
      var riotRequest = new RiotRequest("fake", {
        get: function(region, endpoint, cb) {
          cb(null, {cache: true});
        },
        set: function(region, endpoint, cacheStrategy, data) {
          // jshint unused:false
          throw new Error("Should not be called");
        }
      });

      riotRequest.request('EUW', 'test', '/cacheable', requiredCacheStrategy, function(err, data) {
        assert.ifError(err);
        assert.deepEqual(data, {cache: true});

        done();
      });
    });

    it("should not call the getter function when cache is disabled", function(done) {
      nock('https://euw.api.riotgames.com')
        .get('/cacheable')
        .query(true)
        .reply(200, {ok: true});

      var riotRequest = new RiotRequest("fake", {
        get: function(region, endpoint, cb) {
          // jshint unused:false
          throw new Error("get() should not be called!");
        },
        set: function(region, endpoint, cacheStrategy, data) {
          // jshint unused:false
          throw new Error("set() should not be called!");
        }
      });

      riotRequest.request('EUW', 'test', '/cacheable', false, function(err, data) {
        assert.ifError(err);
        assert.deepEqual(data, {ok: true});

        done();
      });
    });

    it("should use the pre-cache when throttled", function(done) {
      // Set rate limit to 1
      nock('https://euw.api.riotgames.com')
        .get('/throttle')
        .query(true)
        .reply(200, {ok: true}, {
          'x-app-rate-limit-count': '1:1',
          'x-app-rate-limit': '1:1',
          'x-method-rate-limit-count': '1:1',
          'x-method-rate-limit': '1:1'
        });

      // This is delayed to ensure the queue is throttled
      nock('https://euw.api.riotgames.com')
        .get('/pending')
        .query(true)
        .delayBody(1000)
        .reply(200, {ok: true});

      nock('https://euw.api.riotgames.com')
        .get('/cacheable')
        .query(true)
        .reply(200, {ok: true});

      var riotRequest = new RiotRequest("fake", {
        get: function(region, endpoint, cb) {
          if(endpoint === "/cacheable") {
            return cb(null, {cache: true});
          }
          throw new Error("get() should not be called for " + endpoint);
        },
        set: function(region, endpoint, cacheStrategy, data) {
          // jshint unused:false
          throw new Error("set() should not be called!");
        }
      });

      // Throttle the queue
      riotRequest.request('EUW', 'test', '/throttle', function() {
        assert.equal(riotRequest.requestQueues.euwtest.concurrency, 1);
        // This request will take one full second to complete, and since the concurrency is 1, the next request won't start.
        riotRequest.request('EUW', 'test', '/pending', false, function() {});

        setTimeout(function() {
          // And then ensure pre-cache works
          riotRequest.request('EUW', 'test', '/cacheable', true, function(err, data) {
            assert.ifError(err);
            assert.deepEqual(data, {cache: true});

            done();
          });
        }, 10);
      });
    });
  });

  describe("Regions", function() {
    it("should be exposed on riotRequest", function() {
      var riotRequest = new RiotRequest("fake");
      assert.ok(riotRequest.REGIONS.indexOf('euw') !== -1);
    });
  });

  describe("Platforms", function() {
    it("should be available on riotRequest", function() {
      var riotRequest = new RiotRequest("fake");
      assert.ok(riotRequest.getPlatformFromRegion('euw'), 'EUW1');
    });

    it("should be available on riotRequest with any casing", function() {
      var riotRequest = new RiotRequest("fake");
      assert.ok(riotRequest.getPlatformFromRegion('EUW'), 'EUW1');
    });
  });
});


test/setup.js
'use strict';
var nock = require("nock");

beforeEach(function cleanNock() {
  nock.cleanAll();
});


lib/index.js
"use strict";

var async = require("async");
var superagent = require("superagent");
var rarity = require("rarity");
var log = require("debug")("riot-lol-api:request");
var regions = require('./regions.js');


var Agent = require('agentkeepalive').HttpsAgent;
var keepaliveAgent = new Agent();

/**
 * Create a new object for making request to the RiotAPI.
 * Use with new RiotRequest(apiKey, [rate][, cache])
 * Ratelimits will default to a development api key (very slow)
 * Cache must be an object exposing both get() and set() methods
 */
var RiotRequest = function(apiKey, cache) {
  if(!apiKey) {
    throw new Error("Missing Riot API key.");
  }

  if(!cache) {
    log("No cache specified.");
    cache = {
      get: function(region, endpoint, cb) {
        cb(null, null);
      },
      set: function(region, endpoint, cacheStrategy, data) {
        // jshint unused:false
        // Do nothing.
      }
    };
  }

  if(!cache.get || !cache.set) {
    throw new Error("Invalid cache object");
  }

  this.apiKey = apiKey;
  this.cache = cache;

  // Will contain all the queues (one per couple region/method)
  this.requestQueues = {};
};

RiotRequest.prototype.generateQueue = function generateQueue(region, method) {
  var self = this;
  var secondaryLog = require("debug")("riot-lol-api:request:" + region + ":" + method);

  var requestQueue;

  // Assume development key by default, this will be updated on the first request on this queue
  var defaultConcurrency = 20;
  var timeBeforeReset = 1000;

  function readRateLimit(header) {
    return header.split(',').map(v => v.split(':'));
  }

  // Queue worker, loading endpoints from Riot and returning the body
  // This function also handles rate-limiting, retrying after the proper delay.
  // done(err, content, readFromCache)
  function queueWorker(task, done) {
    if(!task.endpoint) {
      return done(new Error("No API endpoint specified"));
    }

    // Strategy for fetching when not in cache
    var fetcher = function getFromRiot(cb) {
      // Do we need to reset concurrency?
      var now = new Date();
      if(now.getTime() - requestQueue.lastNetworkCall.getTime() > timeBeforeReset + 100) {
        requestQueue.concurrency = defaultConcurrency;
        secondaryLog("Resetting queue concurrency to " + defaultConcurrency);
      }
      requestQueue.lastNetworkCall = now;

      secondaryLog("Loading from network " + region + ":" + task.endpoint);

      superagent
        .get("https://" + region + ".api.riotgames.com" + task.endpoint + (task.endpoint.indexOf("?") === -1 ? "?" : "&") + "api_key=" + self.apiKey)
        .agent(keepaliveAgent)
        .timeout(3500)
        .end(function(err, res) {
          if(res && res.headers && res.headers['x-app-rate-limit-count'] && res.headers['x-method-rate-limit']) {
            var appRateCount = readRateLimit(res.headers['x-app-rate-limit-count']);
            var methodRateCount = readRateLimit(res.headers['x-method-rate-limit-count']);

            var appRateLimit = readRateLimit(res.headers['x-app-rate-limit']);
            var methodRateLimit = readRateLimit(res.headers['x-method-rate-limit']);

            var limit = appRateLimit.concat(methodRateLimit);
            var count = appRateCount.concat(methodRateCount);
            var availableConcurrency = Math.max(
              1,
              Math.min.apply(Math, limit.map(function(limit, index) {
                return limit[0] - count[index][0];
              }))
            );

            defaultConcurrency = Math.min.apply(Math, limit.map(l => l[0]));
            timeBeforeReset = Math.min.apply(Math, limit.map(l => l[1]));

            if(requestQueue.concurrency !== availableConcurrency) {
              requestQueue.concurrency = availableConcurrency;
              // secondaryLog("New concurrency for " + region + ": " + availableConcurrency);
            }
          }

          if(err && err.status === 429) {
            // Rate limited :(
            // We'll retry later.
            requestQueue.concurrency = 1;
            var retryAfter = (res.headers['retry-after'] || 2) * 1000;
            secondaryLog("Rate limited, will retry in " + retryAfter + " (pending requests: " + (requestQueue.length() + 1) + ")");

            setTimeout(function() {
              secondaryLog("Restarting after rate limit");
              requestQueue.concurrency = defaultConcurrency;
              queueWorker(task, cb);
            }, retryAfter);
            return;
          }

          if(err && err.timeout) {
            err = new Error("Issues with the Riot API :( [TIMEOUT]");
            err.timeout = true;
          }

          // Mirror actual status code on the error
          if(err) {
            // 500 on Riot side, let's try again just in case this is temporary
            if((err.status === 500 || err.status === 503) && !task.restartedAfter500) {
              task.restartedAfter500 = true;
              secondaryLog("Got a " + err.status + " on " + task.endpoint + ", will try again.");
              setTimeout(function() {
                queueWorker(task, cb);
              }, 25);
              return;
            }

            err.statusCode = err.status;
            err.riotInternal = true;
            err.extra = {
              region: region,
              endpoint: task.endpoint,
              status: err.status,
              currentConcurrency: requestQueue.concurrency,
              defaultConcurrency: defaultConcurrency,
              timeout: err.timeout || false,
              restartedAfter500: task.restartedAfter500
            };
          }


          cb(err, res && res.body, false);
        });
    };

    // Try to read from cache first
    // We potentially already checked in cache once,
    // however we may be in a situation where the same request was queued twice
    // thus resulting in a cache miss.
    // Also, our internal caching layer should be much faster than a real request anyway.
    async.waterfall([
      function getFromCache(cb) {
        if(!task.cacheStrategy) {
          return cb(null, null);
        }

        self.cache.get(region, task.endpoint, rarity.slice(2, cb));
      },
      function actOnCache(cachedData, cb) {
        if(cachedData) {
          secondaryLog("Read from cache " + region + ":" + task.endpoint);
          return cb(null, cachedData, true);
        }

        fetcher(cb);
      }
    ], done);
  }

  // Generate a new queue for this region
  requestQueue = async.queue(queueWorker, defaultConcurrency);
  requestQueue.log = secondaryLog;
  requestQueue.lastNetworkCall = new Date();
  return requestQueue;
};


/**
 * Request a resource from Riot API.
 */
RiotRequest.prototype.request = function request(region, method, endpoint, cacheStrategy, done) {
  region = region.toLowerCase();
  method = method.toLowerCase();
  var queueName = region + method;

  var requestQueues = this.requestQueues;
  var cache = this.cache;

  if(!done) {
    done = cacheStrategy;
    cacheStrategy = false;
  }

  if(!requestQueues[queueName]) {
    // We use one queue per region to manage all calls
    // However, for ease of use and abstraction, we provide a "high-level" function request() which will handle all the queuing process
    // Note though that for this reason, request() can take a long time to process if a lot of queries are already in the region queue.
    // Cached requests are always guaranteed to reply fast however.
    log("Generating new queue for region " + region.toUpperCase() + " and method " + method);
    requestQueues[queueName] = this.generateQueue(region, method);
  }

  var requestQueue = requestQueues[queueName];

  async.waterfall([
    function getFromCache(cb) {
      if(!cacheStrategy || requestQueue.running() < requestQueue.concurrency) {
        // Cache is disabled
        // or concurrency is higher than current job count,
        // which mean there will be a cache query as soon as the task is pushed
        return cb(null, null);
      }
      cache.get(region, endpoint, rarity.slice(2, cb));
    },
    function actOnCache(cachedData, cb) {
      if(cachedData) {
        requestQueue.log("Read from pre-cache " + region + ":" + endpoint + " due to high concurrency");
        return cb(null, cachedData, true);
      }

      requestQueue.push({
        endpoint: endpoint,
        cacheStrategy: cacheStrategy
      }, cb);
    },
    function saveToCache(data, readFromCache, cb) {
      if(!cacheStrategy || readFromCache) {
        return cb(null, data);
      }

      requestQueue.log("Storing in cache " + region + ":" + endpoint + ", strategy " + cacheStrategy);
      // This could theoretically be improved by calling cb() before storing in cache, however doing this in this order ensure that synchronous cache (e.g. lru-cache) will store before we complete, minimizing the risk of cache-miss on future requests.
      cache.set(region, endpoint, cacheStrategy, data);
      cb(null, data);
    }
  ], done);
};


RiotRequest.prototype.REGIONS = regions.REGIONS;
RiotRequest.prototype.getPlatformFromRegion = regions;

module.exports = RiotRequest;
